src/types.rs

```diff
@@ -1,51 +1,154 @@
 //! Concrete node and edge types used in the registry.
 
+extern crate num_traits;
+extern crate quickcheck;
+
+use num_traits::Zero;
+use quickcheck::{Arbitrary, Gen};
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::ops::Add;
+
 /// The type of a node.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub enum NodeType {
     /// A user, eg. contributor, project member etc.
     User,
     /// A project with users as members and contributors.
     Project,
 }
 
+impl Arbitrary for NodeType {
+    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+        if g.next_u32() % 2 == 0 {
+            Self::User
+        } else {
+            Self::Project
+        }
+    }
+}
+
 /// Node data.
-type NodeData = NodeType;
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct NodeData<W> {
+    /// The type for this node.
+    pub node_type: NodeType,
+    /// The total contributions by this user, to *all* projects, if any.
+    pub total_contributions: Option<u32>,
```

Comment:
Hm?

Reply:
This is an attempt to model (dynamic) contributions data for a project or an account, so that we can rely on these when we compute the osrank, see [here](https://github.com/oscoin/osrank-rs/pull/87/files#diff-88175c33d601b2ae4ee5eb28e8f2b549R222-R237). 

As regards why this has been modelled as an `Option` (if your confusion stemmed from that), I am not fully satisfied either. What I wanted to express here is that there are situations where you have no contributions, but at the same time you don't want to use `0`, as that might trigger divisions by 0 down the line and you would have to carefully check for that. However, this now moves the burden of constructing semantically-correct graphs into the registry/graph api layer, as for example having `Some(0)` would be wrong.

I am open to suggestions! Should we simply go with plain `u32`, here?

Reply:
Ah I see - my confusion is partly that not all nodes are users, but the comment says "total contribs by this user". I'd expect this to be part of node_data, with node_type, eg:

```rust
type NodeData = NodeType;
enum NodeType {
  Contributor({ contribs: u32 }),
  Project,
}
```

Does this make sense?

Reply:
> Does this make sense?

At 5PM of a Friday I will say "yes", but I will have to try this in the osrank code to check whether or not there are any blockers. Will give this a go on Monday and report back. "Stay tuned" 😁

Reply:
@cloudhead Ok, after a good restful weekend I can debunk my own statement 😂 : Unfortunately we do need to have a `contribs: Option<u32>` for *both* node types, because in the case of a user that will be (cfr. whitepaper) "the sum of all the contributions this user have done", whereas in the case of a project, this number will be "the sum of all the contributions it received from _all_ accounts".

My suggestion is that we adjust the comment, making it less confusing, but we stick with my original design. What we can reason about though is the use of `Option`, as I have said. Originally I did want to use `Option` to model the fact when a user joins the network it might not have contributed to anything, and ditto for a project, so using `None` would model the "lack of relevant data here". Perhaps using a simple `u32` would work better here (so that we can use `0` as the unit value here), provided we do our due diligence in `osrank` to make sure we always check for division by zero.

Thoughts?

Reply:
:thinking: 
Two things:
1. If we want contributions for both, I would still consider the sum-type, since the meaning of the field is not the same. I think it would be more confusing to have a single thing mean two things in different contexts.
2. I'm not sure why we need the "sum of contributions this user has done"?

Reply:
Yes, the whole point of the sum-type is to not mistakenly mix them up, and ensure that the data is only in scope when the node type is correct (statically).

As for the contribs thing, I'm fairly sure we changed this - I remember talking to @MeBrei and maybe also you about simplifying this logic, since it doesn't make that much sense. I'll try to find any trace of this discussion :thinking:

Reply:
> As for the contribs thing, I'm fairly sure we changed this - I remember talking to @MeBrei and maybe also you about simplifying this logic, since it doesn't make that much sense. I'll try to find any trace of this discussion 🤔

As far as I know (but Merle can debunk me here) what changed is only the `maintain*` weight calculation, which is now equiprobable, whereas for `contrib*` we still need to do the `px/Nx` ratio, and thus you do need the "total number of contributions by x`.

Reply:
`As far as I know (but Merle can debunk me here) what changed is only the maintain* weight calculation, which is now equiprobable, whereas for contrib* we still need to do the px/Nx ratio, and thus you do need the "total number of contributions by x`.`

Yes, for `contrib` and `contrib*` we need the total amount of contributions for the node. There might be ways to work around this, although they have some downfalls:
1) We calculate the total on the fly ie we write a function that goes over each edge of the same type and sums the contributions or after we created the list of contributions, we simply sum it up.
2) We do not use the `total number of ...`. Currently, we use `choose_weighted`, to choose from the `contrib(*)` edges. The function automatically normalizes the weights it chooses from, so it will indirectly devide by the sum of all contributions. The indirectness might make bugs less obvious though.

@adinapoli-mndc am I missing another use of the `total number of ...`?